src/memory.rs

```diff
@@ -54,100 +54,138 @@ impl ByteSliceView {
+    is_nil: bool,
+    ptr: *const u8,
+    len: usize,
 }
 
-// this frees memory we released earlier
-#[no_mangle]
-pub extern "C" fn free_rust(buf: Buffer) {
-    unsafe {
-        let _ = buf.consume();
+impl U8SliceView {
+    pub fn new(source: Option<&[u8]>) -> Self {
+        match source {
+            Some(data) => Self {
+                is_nil: false,
+                ptr: data.as_ptr(),
+                len: data.len(),
+            },
+            None => Self {
+                is_nil: true,
+                ptr: std::ptr::null::<u8>(),
+                len: 0,
+            },
+        }
     }
 }
 
-#[derive(Copy, Clone)]
+/// A Vector type that requires explicit creation and destruction and
+/// can be sent via FFI.
+/// It can be created from `Option<Vec<u8>>` and be converted into `Option<Vec<u8>>`.
+/// This type is always created in Rust and always dropped in Rust.
+/// If Go code wants to consume it's data, it must create a copy and
+/// instruct Rust to destroy it.
 #[repr(C)]
-pub struct Buffer {
-    pub ptr: *mut u8,
-    pub len: usize,
-    pub cap: usize,
+pub struct UnmanagedVector {
+    /// True if and only if this is None/nil. If this is true, the other fields must be ignored.
+    is_nil: bool,
+    ptr: *mut u8,
+    len: usize,
+    cap: usize,
 }
 
-impl Buffer {
-    /// `read` provides a reference to the included data to be parsed or copied elsewhere
-    ///
-    /// # Safety
-    ///
-    /// The caller must make sure that the `Buffer` points to valid and initialized memory
-    pub unsafe fn read(&self) -> Option<&[u8]> {
-        if self.ptr.is_null() {
-            None
-        } else {
-            Some(slice::from_raw_parts(self.ptr, self.len))
+impl UnmanagedVector {
+    /// Consumes this optional vector for manual management.
+    /// This is a zero-copy operation.
+    pub fn new(source: Option<Vec<u8>>) -> Self {
+        match source {
+            Some(data) => {
+                let mut data = mem::ManuallyDrop::new(data);
+                Self {
+                    is_nil: false,
+                    ptr: data.as_mut_ptr(),
+                    len: data.len(),
+                    cap: data.capacity(),
+                }
+            }
+            None => Self {
+                is_nil: true,
+                ptr: std::ptr::null_mut::<u8>(),
+                len: 0,
+                cap: 0,
+            },
         }
     }
 
-    /// consume must only be used on memory previously released by from_vec
-    /// when the Vec is out of scope, it will deallocate the memory previously referenced by Buffer
-    ///
-    /// # Safety
-    ///
-    /// if not empty, `ptr` must be a valid memory reference, which was previously
-    /// created by `from_vec`. You may not consume a slice twice.
-    /// Otherwise you risk double free panics
-    pub unsafe fn consume(self) -> Vec<u8> {
-        if self.ptr.is_null() {
-            vec![]
+    pub fn len(&self) -> Option<usize> {
```

Comment:
Wouldn't this be easier to work with if it returns `usize` (like `len` normally does).
The nil case can return 0, and `is_none`, as well as consuming will tell the difference.
Also, typically they define `is_empty()` which is basically `len() == 0`

Reply:
No because if this is None/nil, we must not look into the other field. They are allowed to contain garbage. E.g.

```rust
UnmanagedVector {
    is_nil: true,
    ptr: std::ptr::null_mut::<u8>(),
    len: 20,
    cap: 4,
}
```

is valid.

If in your call you want to treat None and zero length the same, you just do `.len().unwrap_or_default()`.

Reply:
I see your point. Maybe you could use another name (as `len()` has the same signature everywhere). But no big deal here, just an API surprise (due to common function name)

Reply:
Right. I think the issue here is that `UnmanagedVector` is a C-compatible `Option<Vec<u8>>`. So we can only add the obvious `Option` methods directly. If we do some kind of `.unwrap_ref().len()`, we would allow reading data without consuming it.

But most likely we can get rid of that completely as it is only used in tests. In other cases we can just `consume()` and then inspect.

Reply:
Okey, removed. It was not used.

Also I now made clear that both `UnmanagedVector`/`U8SliceView` represent Rust `Option`s by using "none" instead of "nil" consistently.